src/scheduler/MpiWorld.cpp

```diff
@@ -25,40 +30,109 @@ MpiWorld::MpiWorld()
   , cartProcsPerDim(2)
 {}
 
-faabric::scheduler::FunctionCallClient& MpiWorld::getFunctionCallClient(
-  const std::string& otherHost)
+void MpiWorld::sendRemoteMpiMessage(
+  int sendRank,
+  int recvRank,
+  const std::shared_ptr<faabric::MPIMessage>& msg)
 {
-    auto it = functionCallClients.find(otherHost);
-    if (it == functionCallClients.end()) {
-        // The second argument is forwarded to the client's constructor
-        auto _it = functionCallClients.try_emplace(otherHost, otherHost);
-        if (!_it.second) {
-            throw std::runtime_error("Error inserting remote endpoint");
+    // Assert the ranks are sane
+    assert(0 <= sendRank && sendRank < size);
+    assert(0 <= recvRank && recvRank < size);
+
+    // Initialise the endpoint vector if not initialised
+    if (mpiMessageEndpoints.size() == 0) {
+        for (int i = 0; i < size * size; i++) {
+            mpiMessageEndpoints.emplace_back(nullptr);
         }
-        it = _it.first;
     }
-    return it->second;
+
+    // Get the index for the rank-host pair
+    int index = getIndexForRanks(sendRank, recvRank);
+    assert(index >= 0 && index < size * size);
+
+    // Lazily initialise send endpoints
+    if (mpiMessageEndpoints[index] == nullptr) {
+        // Get host for recv rank
+        std::string host = getHostForRank(recvRank);
+        assert(!host.empty());
+        assert(host != thisHost);
+
+        // Get port for send-recv pair
+        int port = getMpiPort(sendRank, recvRank);
+        // TODO - finer-grained checking when multi-tenant port scheme
+        assert(port > 0);
```

Comment:
This assertion should live in `getMpiPort` if we want to keep it, but it seems like it will only fail if the ranks are invalid, which is probably something we should check higher up the stack with an exception and error message.